addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/api/manager/MiHomeApiManagerImpl.java

```diff
@@ -0,0 +1,204 @@
+
+    }
+
+    @Override
+    public JsonObject unregisterGateway(int id) {
+        Map<String, Object> parameters = new HashMap<String, Object>();
+        parameters.put(DeviceConstants.DEVICE_ID_KEY, id);
+
+        return execute(CONTROLLER_DEVICES, ACTION_DELETE, parameters);
+    }
+
+    @Override
+    public JsonObject upgradeGatewayFirmware(int id) {
+        Map<String, Object> parameters = new HashMap<String, Object>();
+        parameters.put(DeviceConstants.DEVICE_ID_KEY, id);
+
+        return execute(CONTROLLER_DEVICES, ACTION_UPDATE_FIRMWARE, parameters);
+    }
+
+    @Override
+    public JsonObject listSubdevices() {
+        return execute(CONTROLLER_SUBDEVICES, ACTION_LIST);
+    }
+
+    @Override
+    public JsonObject registerSubdevice(int gatewayID, String deviceType) {
+        Map<String, Object> parameters = new HashMap<String, Object>();
+        parameters.put(DeviceConstants.SUBDEVICE_PARENT_ID_KEY, gatewayID);
+        parameters.put(DeviceConstants.DEVICE_TYPE_KEY, deviceType);
+        return execute(CONTROLLER_SUBDEVICES, ACTION_CREATE, parameters);
+
+    }
+
+    @Override
+    public JsonObject showSubdeviceInfo(int id) {
+        Map<String, Object> parameters = new HashMap<String, Object>();
+        parameters.put(DeviceConstants.DEVICE_ID_KEY, id);
+        // With this parameter we exclude the historical usage data ("1" - to include, "0" - to exclude)
+        // The historical data is from no importance to the user,
+        // it consists of the number of events for each hour from the last 24 hours.
+        parameters.put(DeviceConstants.SUBDEVICE_INCLUDE_USAGE_DATA, 0);
+        return execute(CONTROLLER_SUBDEVICES, ACTION_SHOW, parameters);
+    }
+
+    @Override
+    public JsonObject updateSubdevice(int id, String label) {
+        Map<String, Object> parameters = new HashMap<String, Object>();
+        parameters.put(DeviceConstants.DEVICE_ID_KEY, id);
+        parameters.put(DeviceConstants.DEVICE_LABEL_KEY, label);
+        return execute(CONTROLLER_SUBDEVICES, ACTION_UPDATE, parameters);
+    }
+
+    @Override
+    public JsonObject unregisterSubdevice(int id) {
+        Map<String, Object> parameters = new HashMap<String, Object>();
+        parameters.put(DeviceConstants.DEVICE_ID_KEY, id);
+        return execute(CONTROLLER_SUBDEVICES, ACTION_DELETE, parameters);
+    }
+
+    private JsonObject execute(String controller, String action) {
+        return execute(controller, action, new HashMap<String, Object>());
+    }
+
+    private JsonObject execute(String controller, String action, Map<String, Object> content) {
+
+        Properties httpHeaders = new Properties();
+
+        // Set basic authentication header
+        String authData = configuration.getUserName() + ":" + configuration.getPassword();
+        String basicAuthentication = "Basic " + Base64.getEncoder().encodeToString(authData.getBytes());
+        basicAuthentication = basicAuthentication.trim();
```

Comment:
I prefer to keep it, I placed it for a reason, but I am not 100 percent sure why. (a comment would be precious now ..)
It could be that before I used the Apache Base64.encodeBase64String() that used to [append a new line ](http://stackoverflow.com/questions/20040539/new-line-appending-on-my-encrypted-string) at the end of the header.